3단계 - 기능 우선 패키지 구성하기

seungjoopet opened the pull request:
- 기능 중심으로 패키지를 분할함
- 메뉴/상품/배달주문/포장주문/매장주문/매장테이블 으로 분할함
- 비즈니스 없이 외부 API를 호출하기만 하는 client를 둠

안녕하세요 리뷰어님!
`계층 우선 패키지 구조를 BOUNDED CONTEXT 중심의 패키지 구조로 리팩터링한다.` 요구사항을 제대로 적용한지 잘 모르겠습니다^_ㅠ 어렵네욥.. 

1. `변경의 단위`로 컨텍스트를 분리하였고, 기존에 RestaurantTable로 명명했던 것을 EatinTable로 변경하였습니다.
    * 매장주문 하위에 매장테이블을 변경의 단위가 다르다고 생각하여 분리하긴 했는데, 결국 매장테이블도 매장주문의 구성요소 인 것 같아서 여전히 고민인 부분입니다.
2. `PurgomalumClient`의 경우 몇몇 컨텍스트에서 공통으로 쓰이므로 비즈니스를 하나도 갖지 않고, remote를 호출해주기만 하는 client/ 라는 패키지를 하나 만들고 거기에 배치하였습니다. 
이렇게 만든 client는 필요한 곳에서 자신의 컨텍스트에 맞게 호출하여 사용할 수 있도록 구성했습니다(ex. `ProductPurgoalmumChecker`, `MenuPurgoalmumChecker`)

Commit:
feat(movePackage) 새 구성을 위한 기존 코드 패키지 이동

Commit:
feat(step3) 3단계 - 기능 우선 패키지 구성하기

- 기능 중심으로 패키지를 분할함
- 메뉴/상품/배달주문/포장주문/매장주문 으로 분할함
- 비즈니스 없이 외부 API를 호출하기만 하는 client를 둠

Commit:
feat(eatinTable) restaurantTable -> eatinTable로 변경 및 패키지 분리

sah3122 commented on src/main/java/kitchenpos/client/purgomalum/DefaultPurgomalumClient.java:
```diff
@@ -0,0 +1,27 @@
+import org.springframework.web.util.UriComponentsBuilder;
+
+@Component
+public class DefaultPurgomalumClient implements PurgomalumClient{
```
`DefaultPurgomalumClient`기능을 컨텍스트에 포함하지 않고 root 패키지에 두어주셨네요. 👍 
이와 같은 기능이 많아지는 경우 root 패키지가 커질수 있고 특정 컨텍스트에서 요구하는 기능들이 추가되는 경우 어려움이 있을수 있습니다. 
반면 코드 중복을 제거할수 있다는 장점이 있어요. 
두 방식에 대한 장단점을 고민해보시면 좋을것 같습니다 😄

seungjoopet replied:
안그래도 이 부분에서 많은 고민을 했는데요!

`product`의 `비속어`와, `menu`의 `비속어`는 다른 개념일 수 있기 때문에, 
지금 root 패키지에 있는 `PurgomalumClient`는 아무런 비즈니스도 갖지 않고, remoteAPI를 호출하는 일만 수행해줄 예정입니다.
(원격API warpper 수준으로요)

결국 `PurgomalumClient` 는, 원격 클라이언트 호출 정도, 구현 레벨로 가면 restTemplate.call() 정도인데요.
이 코드를 `product`와 `menu`가 모두 가지고 있는 것과 비교했을 때 클라이언트를 따로 두는게 유의미할까? 싶었습니다. 

`PurgomalumClient`를 사용하는 곳이 3~4개 되어서 같은 client코드를 사용하는 곳이 많아진다면 그 땐 중복 줄이는게 의미가 있을 수 있겠지만,
또 한편으로는
* `A` 컨텍스트에서 사용하는 `PurgomalumClient` 와 `B` 컨텍스트에서 사용하는 `PurgomalumClient` 가 
다른 httpClient 구현체를 이용해줘야 한다면? 
* `Purgomalum리모트API` 가 컨텍스트별로 다른 헤더값을 요구한다면? 그리고 그 헤더값을 바탕으로 response model을 조정한다고 하면?

그럼 결국 `PurgomalumClient` 는 특정 컨텍스트에 종속적인 값이나 상태를 가지게 될거고 결국 `아무런 비즈니스도 갖지 않고, remoteAPI를 호출하는 일만 수행하여 중복을 줄인다`는 원래의 의도를 지키지 못할 것 같습니다.

뭐 이런저런 많은 생각들을 했습니다..ㅎㅎ 
그래서 결론은 client를 따로 공통으로 분리하는게 유의미한 순간이 있을까? 싶었어요.
리뷰어님은 혹시 어떻게 생각하실까요~?

sah3122 replied:
저도 비슷한 생각을 가지고 있습니다 😄 
client 를 분리하는것이 유의미한 순간이 온다면 분리할것이고 그전까지는 공통으로 사용하는것이 경험상 좋았던건 같아요 !

sah3122 commented on src/main/java/kitchenpos/deliveryorder/adapter/in/DeliveryOrderRestController.java:
```diff
@@ -0,0 +1,10 @@
+package kitchenpos.deliveryorder.adapter.in;
```
port & adapter 구조를 반영해주셨네요 👍 
주문 컨텍스트를 주문 타입에 따라 분리해주셨는데 아래와 같이 주문이라는 큰 컨텍스트와 타입에 따른 하위 컨텍스트로 변경해보는건 어떻게 생각하시나요 ?
```
- order
  - deliveryorder
  - takeorder
  - eatinorder
```
위와 같은 구조를 가지고 갔을때 주문에서 필요하는 공통 기능을 order 패키지 하위에 둘수도 있습니다 😄

sah3122 commented on src/main/java/kitchenpos/deliveryorder/domain/DeliveryOrderLineItem.java:
```diff
@@ -0,0 +1,5 @@
+package kitchenpos.deliveryorder.domain;
+
+public class DeliveryOrderLineItem {
```
OrderLineItem을 주문타입에 따라 분리해주셨네요 👍 
주문 항목은 주문 컨텍스트에서 공통적으로 사용하는 도메인 입니다. 이처럼 분리하는것과 공통으로 사용하는 것에 대한 차이점을 고민해보세요 😄

sah3122 commented on src/main/java/kitchenpos/menu/adapter/in/MenuGroupRestController.java:
```diff
@@ -0,0 +1,20 @@
+package kitchenpos.menu.adapter.in;
```
Menu 컨텍스트 하위에 MenuGroup 도메인이 들어가는건 자연스러운것 같습니다. 👍 
한가지 고민거리로 컨텍스트 내부에서 Aggregate을 기준으로 패키지를 분리하는것이 어떤지 고민해보시면 좋을것 같아요.

seungjoopet replied:
말씀 주신 부분이 잘 이해가 안가서, 좀만 더 풀어서 말씀해주실 수 있으실까요~?

`컨텍스트 내부에서 Aggregate을 기준으로 패키지를 분리하는것`

Menu컨텍스트 안에, menu와 menuGroup이 있다고 생각하는데,
menu와 menuGroup은 일관성을 유지해야하는 하나의 단위가 아니기 때문에
Menu 라는 컨텍스트 하위에 
menu, menuGroup 이 있으면 좋을 것 같다는 말씀이실까요?

sah3122 replied:
네 맞습니다 ! 
그렇게 분리하게된다면 menu 패키지 하위에 menu 라는 패키지가 생기게 됩니다. 
이런 경우 명시적으로 Menu 와 MenuGroup을 포함할수 있는 컨텍스트를 정의하는것도 좋은 방법이라고 생각합니다 😄

sah3122 commented on src/main/java/kitchenpos/deliveryorder/domain/DeliveryOrder.java:
```diff
@@ -0,0 +1,4 @@
+package kitchenpos.deliveryorder.domain;
```
repository와 같은 기능들은 어디에 위치 하게 될까요 ?

seungjoopet replied:
음 `repository` 인터페이스의 위치가 좀 고민스러웠는데요! (여전히 고민중)

`repository`는 도메인 영속화의 명세이기 때문에 domain entity layer(domain/)에 넣어야 할까 생각했는데요,
포트앤어댑터 패턴을 보면 domain의 repository interface또한 out port에 있더라구요.
(포트앤 어댑터 패턴을 제대로 써본 적이 없고, 이번에 리뷰 많이 해주실 때 한번 실습해보려고 하는 것이라 잘 모릅니당ㅎ_ㅎ)

<img width="417" alt="스크린샷 2023-09-10 오전 11 16 19" src="https://github.com/next-step/ddd-strategic-design/assets/19169557/066a3e2e-f029-4ca2-b120-da4d31c934f1">
<img width="412" alt="스크린샷 2023-09-10 오전 11 16 53" src="https://github.com/next-step/ddd-strategic-design/assets/19169557/3ec8f71f-5a26-49de-a92c-29b6b25b3d2e">

> 제가 참고한 부분 한번 첨부해봅니당😀



그래서 포트앤어댑터 관련 아티클들을 몇개 찾아봤는데, 
`repository`를 domain/에 둔 곳도 있고, port/에 둔 곳도 있더라구요.
패키지 구조(이름)만 조금씩 다를 뿐, 모두 domain entity를 향하여 의존성을 가진 것 같아서 어느 패키지에 위치하느냐 보다 `도메인엔티티를 잘 지켜주느냐`가 핵심인 것 같다는 생각했습니다. 
그래서 설계하는 제 입장에서는
`repository`를 domain/에 둬도 되고, port/out에 둬도 되는데요. 
패키지 구조와 네이밍 자체가 port&adapter 아키텍쳐임을 소리치고 있기 때문에
새로운 사람이 이 프로젝트를 접해야 할 때 `repository` 가 domain/ 안에 있으면 의아해하지 않을까? 싶었습니다.

그래서 일단 `repository`는 port/out에 위치시켰습니다. 

위에 말한 것들중 제가 잘못이해하고 있는 것들이 있거나 다른 의견이 있으시면 편하게 말씀주세요!! :)

sah3122 replied:
port & adapter 아키텍처를 차용한다면 repository 의 구현체는 변경될수 있기 때문에 out 에 위치하는것이 올바르다고 생각합니다. 
조금더 나아간다면 Entity 를 Aggregate으로 사용(JPA Entity 와 Domain Class를 분리하지 않음)한다면  Entity 가 특정 Repository 구현과 결합이 발생하기 때문에 port & adapter 아키텍처의 구조를 완벽하게  반영하기 어려운 구조가 됩니다. 
이런 경우에는 Entity 와 Aggregate 객체를 분리할수도 있어요 😄

seungjoopet replied:
>Entity 를 Aggregate으로 사용(JPA Entity 와 Domain Class를 분리하지 않음)한다면 Entity 가 특정 Repository 구현과 결합이 발생하기 때문에 port & adapter 아키텍처의 구조를 완벽하게 반영하기 어려운 구조

안그래도 JPA를 써본적은 없지만 익히 들어와서 JPA가 DomainClass와 매핑되게 쓴다 & 그래서 특정 기술이 도메인레이어로 침투하게 된다. 라는 건 알고 있었습니다. JPA가 주는 이점이 매우 크기 때문에 침투하더라도 JPA와 Domain Entity를 굳이 분리하진 않는다. 까진 들었는데, 
혹시 이를 분리해서 얻을 수 있는 장점이 있을까요? 혹은 관련 아티클이 있으면 추천주시면 감사하겠습니다 :)

저도 조금만 더 찾아보고 이번 실습 때 기회가 된다면 Entity와 Aggregate를 분리하는 방향도 한번 검토해보겠습니다 :)

sah3122 replied:
제가 생각하는 분리해서 얻는 장점으론 아래과 같습니다. 
1. 특정 구현체에 종속하지 않는다.
2. Database와의 결합을 제거하여 도메인 구조를 유연하게 가져갈수 있다.

더 많은 장점이 있을것 같지만 단점도 명확합니다.
코멘트 남겨주신것 처럼 JPA의 기능을 못쓰게 되는건 장점이자 단점일것 같네요 ㅎㅎ 
추천해드릴 아티클은 없는것 같아요 😢

sah3122 reviewed (changes requested):
기능우선 패키지 정의 잘해주셨습니다 👍 
몇가지 코멘트 남겨두었는데 확인 부탁드러요 !

Commit:
refactor(review) 리뷰 일부 반영

1. Menu(음식&메뉴) 라는 컨텍스트 하위에 Menu, MenuGroup, Product 도메인 들어감.
    * 키친포스에서 음식(Product)는 메뉴를 구성하는 것이므로 따로 물리적 경계를 구분하지 않고 Menu 안으로 경계를 넓게 가져갈까 합니다.
2. 각 주문들은 `주문`이라는 카테고리는 같고 비록 공통적인 부분도 있지만, 변경의 사유와 비즈니스 규칙이 모두 다르기 때문에 DeliveryOrder, TakeoutOrder, EatinOrder 컨텍스트로 구분

seungjoopet commented:
후아 어렵네요. 잘 모르다보니 생각이 시시각각 바뀝니다 ^_ㅠ 

지금 생각으로는,
* 바운디드컨텍스트라는 것은 `문제해결영역`이고, 지금 우리의 문제는 `키친포스` -> `상품을 등록하여 메뉴를 생성하고 각 주문을 알맞은 방식으로 처리`하는 것 
* 키친포스의 유연하게 바뀌어야 하는 `핵심 하위 도메인`은 각 `주문`임

그러므로, 
1. Menu(음식&메뉴) 라는 컨텍스트 하위에 Menu, MenuGroup, Product 도메인 들어감. 
    * 키친포스에서 음식(Product)는 메뉴를 구성하는 것이므로 따로 물리적 경계(헥사고날 분리)를 구분하지 않고 Menu 안으로 경계를 넓게 가져갈까 합니다.
2. 각 주문들은 `주문`이라는 카테고리는 같고 비록 공통적인 부분도 있지만, 변경의 사유와 비즈니스 규칙이 모두 다르기 때문에 DeliveryOrder, TakeoutOrder, EatinOrder 컨텍스트로 구분
    * `eatinTable`도 Product와 같은 맥락으로 기존에 구분한 물리적경계를 허물고 `eatinOrder`의 안으로 넣었습니다. (키친포스에 매장기물관리 기능이 추가된다든지, 매장테이블이 가지는 비즈니스가 더 식별된다면 그 때서야 경계가 분리되지 않을까 합니다)

위와 같은 생각으로 패키지 구조를 변경해보았습니다.



<img width="896" alt="context-1" src="https://github.com/next-step/ddd-strategic-design/assets/19169557/f09140f7-140d-4712-b375-9647904331c4">

sah3122 commented on src/main/java/kitchenpos/menu/adapter/in/ProductRestController.java:
```diff
@@ -0,0 +1,10 @@
+package kitchenpos.menu.adapter.in;
```
Product 컨텍스트가 Menu 컨텍스트에 포함되었다는것을 명시적으로 알려주기 위해 menu.product.adpater 와 같은 구조로 나타내는것이 더 직관적일수 있습니다. 
개인적으론 컨텍스트의 경계를 패키지로 나타내어 컨텍스트간 결합이 발생했는지 import 구문으로 확인하곤 합니다 !

sah3122 commented:
https://github.com/next-step/ddd-strategic-design/pull/320#issuecomment-1712714860
코멘트에도 남겨두었는데 컨텍스트의 경계를 패키지로 나타내는것과 상위 컨텍스트로 묶어보는것도 경험해보면 종흘것 같습니다 😄

sah3122 reviewed (changes requested):
피드백 반영 잘해주셨습니다 👍 
남겨주신 코멘트에 대한 답변 추가해두었는데 확인 부탁드리며 다음 요청시 머지하도록 하겠습니다 🙏

Commit:
refator(review) 상위 컨텍스트 추가

- 명시적인 컨텍스트를 표현하기 위해 상위 컨텍스트를 추가함
- 메뉴 : 메뉴, 메뉴그룹, 음식 / 배달음식 / 포장음식 / 매장음식 : 매장음식, 매장테이블

seungjoopet commented:
global한 client를 제거하였습니다.
그리고, 상위 컨텍스트가 있는 곳에 상위컨텍스트 개념을 부여해보았습니다.
부여하니 제가 그린 그림대로 패키지가 나뉘게 된 것 같아요 :)

한가지 고민이 들었는데, 
상위컨텍스트를 나타내고 하위를 다른 패키지로 분리하니 헥사고날 안에 작은 헥사고날들이 또 들어가게 된거라 이 작은 헥사고날끼리 어디까지 참조하게 하게 만들지가 고민이 들었습니다.

이건 한번 구현해보면서 더 공부하고 고민하고 보강해보겠습니다!

sah3122 commented on src/main/java/kitchenpos/eatinorder/eatinorder/adapter/in/EatinOrderRestController.java:
```diff
@@ -1,4 +1,4 @@
-package kitchenpos.eatinorder.adapter.in;
+package kitchenpos.eatinorder.eatinorder.adapter.in;
```
`eatinorder` 하위에 `eatinorder`라는 컨텍스트가 중복되는것이 어색할수 있습니다.
매장 식사와 매장 테이블을 포함하는 상위 컨텍스트를 도출해보세요 😄 
`menu`에도 적용해보면 좋을것 같아요 !

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
적지않은 피드백들 반영해주신다고 고생많으셨어요 🙏 
남아있는 미션들도 끝까지 완주하시길 바랄게요 !

sah3122 commented:
`상위컨텍스트를 나타내고 하위를 다른 패키지로 분리하니 헥사고날 안에 작은 헥사고날들이 또 들어가게 된거라 이 작은 헥사고날끼리 어디까지 참조하게 하게 만들지가 고민이 들었습니다.`
컨텍스트의 경계가 정해졌으니 참조를 제거해야한다고 생각합니다. 
직점 참조할수 있는 부분은 Aggregate Root 로 연결되어 있는 Child Aggregate 정도가 될것 같습니다 😄 
직접 참조를 하지않고 Domain Service / Event 를 활용해보세요 !

seungjoopet commented:
@sah3122 리뷰어님, 
https://github.com/next-step/ddd-strategic-design/pull/320/commits/f2de9ae02ae0e07f3a2a1c01d753448e1cfec9d1

요고 반영하고 3단계를 진행하려고 보니, 

https://github.com/next-step/ddd-strategic-design/pull/320#issuecomment-1713501018
여기에 남긴 `헥사고날 안에 작은 헥사고날들이 또 들어가게 된거` 의문이 좀 더 커지더라구요!

제가 설계한 헥사고날은 BoundedContext의 public interface(즉, usecase) 문제 해결을 위한 Usecase 단위이고, 
이를 달성하기 위한 domain layer는 aggregate는 분리되어 있지만 서로 간접 참조를 할 수 있게 구성했는데요.

하위 컨텍스트 별로 다시 헥사고날을 만들게 되니, 
domain별로 헥사고날이 나뉘어지는 구조가 되었습니다. 그래서 내부를 왜 더 쪼개야하지? 이런 의문이 들었습니다. 

리뷰어님께서는 헥사고날의 구분 단위를 어떻게 보시는 걸까요~?

<img width="760" alt="구성질문" src="https://github.com/next-step/ddd-strategic-design/assets/19169557/00041773-2631-444b-bba0-c990c71fbcf3">



aggregate를 명시적으로 구분하는게 목표라면, 저는 아래처럼 가는게 좀 더 적절하지 않을까? 싶었습니다~!

아래는 menu/ 의 패키지구조입니다. 

menu/
* adapter
    * in
    * out
* application
    * port
        * in
        * out  
* domain
    * menu
    * menugroup
    * product

sah3122 commented:
핵사고날의 구분이 잘 이해 되지 않습니다 ㅎㅎ ...  
비즈니스 경계를 기준으로 구분하는것을 질문주신것 같아 답변 드려요. 

바운디드 컨텍스트 내부에는 여러 비즈니스 모델(도메인)이 존재합니다. 
각 도메인은 하나의 Aggreagate으로 정의될수도, 여러 Aggregate으로 정의될수도 있습니다. 바운디드 컨텍스트가 하나의 Aggregate으로도 정의될수 있죠. 

여기서 정의할수 있는건 비즈니스 경계를 기준으로 패키지를 정의할수 있을것 같아요. 

비즈니스 경계를 본다면 같이 menu(menuGroup 포함) / product 로 나눌수도 menu / menuGroup / product 로 나눌수도, 세가지 도메인 모두 menu 로 정의할수 있습니다. 
menu 와 menuGroup 은 하나의 비즈니스로 연결되어 있지만 product 는 menu의 구성 요소이자 별도의 라이프사이클을 가지기 때문에 
개인적으론 menu(menuGroup 포함) / product 가 좀더 자연스러운 경계라고 볼수있을것 같아요. 

코멘트의 예시처럼 패키지를 정의하면 adapter / application 내부에서의 경계를 나타낼수 없습니다. 
경계를 정의하기 위해 하위 패키지들도 헥사고날 아키텍처를 가져가며 특정 패키지가 커지는것을 방지하기 위함이라고 생각해주시면 될것 같아요. 

패키지를 나누는것에 대해 몇가지 아티클 소개드릴게요 ! 
https://ssowonny.medium.com/package-by-feature-in-clean-architecture-projects-e14d25e3905e
https://medium.com/sahibinden-technology/package-by-layer-vs-package-by-feature-7e89cde2ae3a
http://www.javapractices.com/topic/TopicAction.do?Id=205

seungjoopet commented:
추후 참고를 위해 대화 내용 남겨놓겠습니다 :)

<img width="1131" alt="스크린샷 2023-09-12 오후 12 58 00" src="https://github.com/next-step/ddd-strategic-design/assets/19169557/87344609-c328-4af4-ba1e-7a10f3882cf3">
